WoofWare.NUnitTestRunner.Lib/ParallelQueue.fs

```diff
@@ -0,0 +1,124 @@
+/// TODO: consume the cancellation token
+type ParallelQueue
+    (_parallelism : int option, _scope : Parallelizable<AssemblyParallelScope> option, ?ct : CancellationToken)
+    =
+    let rec processTask (state : MailboxState) (m : MailboxProcessor<MailboxMessage>) =
+        async {
+            let! message = m.Receive ()
+
+            match message with
+            | Quit rc -> rc.Reply ()
+            | BeginTestFixture (tf, rc) ->
+                match state with
+                | Running (current, rest) ->
+                    let state = Running (current, (tf, rc) :: rest)
+                    return! processTask state m
+                | Idle ->
+                    let state = Running (tf, [])
+                    rc.Reply (TestFixtureRunningToken tf)
+                    return! processTask state m
+            | EndTestFixture (TestFixtureRunningToken tf, rc) ->
+                match state with
+                | Idle ->
+                    return failwith "Caller has somehow called EndTestFixture while we're not running a test fixture"
+                | Running (current, rest) ->
+                    if not (Object.ReferenceEquals (current, tf)) then
+                        return
+                            failwith
+                                "Caller has somehow called EndTestFixture while we're not running that test fixture"
+
+                    rc.Reply ()
+
+                    match rest with
+                    | [] -> return! processTask Idle m
+                    | (head, rc) :: tail ->
+                        rc.Reply (TestFixtureRunningToken head)
+                        return! processTask (Running (head, tail)) m
+            | RunTest message ->
+                // Currently we rely on the caller to only send this message when we've given them permission through
+                // the StartTestFixture method returning.
+                { new ThunkEvaluator<_> with
+                    member _.Eval t rc =
+                        use ec = ExecutionContext.Capture ()
+                        ExecutionContext.Run (ec, (fun _ -> rc.Reply (t ())), ())
+                        FakeUnit
+                }
+                |> message.Apply
+                |> function
+                    | FakeUnit -> ()
+
+                return! processTask state m
+        }
+
+    let mb = new MailboxProcessor<_> (processTask MailboxState.Idle)
+    do mb.Start ()
+
+    /// Request to run the given action on its own, not in parallel with anything else.
+    /// The resulting Task will return when the action has completed.
+    member _.NonParallel<'a> (action : unit -> 'a) : 'a Task =
+        ThunkCrate.make action >> RunTest |> mb.PostAndAsyncReply |> Async.StartAsTask
+
+    /// Request to run the given action, freely in parallel with other running tests.
+    /// The resulting Task will return when the action has completed.
+    member _.Parallel<'a> (action : unit -> 'a) : 'a Task =
```

Comment:
Unclear if this is the right interface. Also it probably should take a `TestFixtureRunningToken`.